### ISSUE TITLE
Persistent storage / replicated volume recommendation for docker swarm mode in 2020

### ISSUE THREAD

**Issue by OCram85** [tiangolo/dockerswarm.rocks#50]
Hi,
It would be awesome to have a guide or a recommendation what kind of solutions could be used to set up replicated volumes.
I recently looked into solutions like [GlusterFS](https://www.gluster.org/), [Ceph](https://ceph.io/),  [REX-Ray](https://github.com/rexray/rexray)...

**Comment by tiangolo** [comment]
Yeah, the thing is that I don't have a recommendation for that. I don't think any of the currently available tools solve all the related problems.

For example, you probably wouldn't put the files of a SQL DB in one of those systems. 

They would work best for binary files stored as blobs that can be retrieved and saved, etc. But not constantly updated, indexed, etc. And in those cases, I would think [MinIO](https://min.io/) could work better.

**Comment by OCram85** [comment]
You're absolutely right.
I think it would be helpful to compare some current solutions with your mentioned use cases. It seems to be a good starting point for new docker swarm users which are looking for best practice implementation. As a beginner it's quite hard to find a simple route in 2020 for replicated volumes.

Maybe it would also help to list the mostly used products in 2020 and describe some implemtations for some use cases?

**Comment by ramonck** [comment]
Hey,

Maybe Galera is a good approach? I have not tested just found it in my searches. Insteresting because of the realtime round-robin thing so if falls in real-time another DB takes over due to the realtime replication no loss, sounds interesting. Not sure how it works in practices.
https://blog.ruanbekker.com/blog/2019/05/10/running-a-ha-mysql-galera-cluster-on-docker-swarm/

This looks like a paid product and they say to have a free version: https://severalnines.com/product/clustercontrol

**Comment by ramonck** [comment]
https://docs.docker.com/engine/examples/couchdb_data_volumes/

**Comment by MichaelLeeHobbs** [comment]
We use swarm in production with 80+ containers. In our environment ie AWS we use EFS for persistence. We have EFS mounted to all the swarm nodes and the docker files map the volumes to the efs mount. Works well for the small amount of persistence we need. EFS is expensive so if you had a large volume of data that needed persistence I'd use something else. As a general note, we will be moving to k8s so I'm not sure what we will do in the future. Overall, some sort of NFS mount has worked well in my experience. It's also simple to use once setup.

**Comment by pascalandy** [comment]
I use Resilio. But I'm about to switch [to this](https://trello.com/c/uwUD6H67/505-utility-syncthing-to-replace-resilio).

**Comment by Ivanello** [comment]
guys, what can You say about this option?
https://github.com/vieux/docker-volume-sshfs

**Comment by yolombana** [comment]
Hola yo utilizo una infraestructura de almacenamiento distribuido con glusterfs replicado en 6 maquinas generando 6 volúmenes replicados con un quórum del 48% y funcionado en un cluster de docker swarm. Funciona bien. Solo debes tener cuidado con el cerebro dividido

**Comment by decentral1se** [comment]
https://github.com/moby/moby/issues/39624

**Comment by tiangolo** [comment]
I should let you know, that I had to deprecate this website and ideas, I would no longer recommend Docker Swarm Mode for new projects: https://dockerswarm.rocks/swarm-or-kubernetes/ 🥲

**Comment by OCram85** [comment]
Thanks for the update on this. 

Btw, I decided using NFS for volumes and still using docker swarm in production ;).

**Comment by thebigbone** [comment]
> Thanks for the update on this.
> 
> Btw, I decided using NFS for volumes and still using docker swarm in production ;).

Hello, I know it's a very old issue but how do you handle the syncing issue with NFS? Do you run NFS container on all the nodes, kinda like a daemonset?

**Comment by OCram85** [comment]
It depends on the environments I'm using:

- Currently I use either an external NFS Server (SPOF) if I don't need to fulfill a HA SLA... 
- Or I'm using again NFS exports provided by a Netapp Cluster / SAN infrastructure.

But I'm also currently testing [seaweedFS](https://github.com/seaweedfs/seaweedfs) as a lightweight ceph alternative. See: https://github.com/seaweedfs/seaweedfs/wiki/SeaweedFS-in-Docker-Swarm

**Comment by thebigbone** [comment]
I see. Does a simple setup such as mounting NFS on every node and using a shared folder work?

**Comment by jloehel** [comment]
A pity that this issue was closed. Still using docker swarm for production, too. I usually use GlusterFS with the glusterfs docker plugin to create docker (glusterfs) volumes. Before I used NFS. I had one docker export where I stored a all the docker volumes. Simply declared like that:
~~~
    volumes:
      my_nfs_volume:
        driver: local
        driver_opts:
          type: nfs
          o: "addr=[NFS_SERVER_IP],rw,nfsvers=4"
          device: ":[PATH_ON_NFS_SERVER]"
~~~

For a ha NFS setup I suggest pacemaker and drbd: https://documentation.suse.com/sle-ha/15-SP3/single-html/SLE-HA-nfs-storage/. I am pretty sure not everyone has the money for a Netapp. ^^
